### DISCUSSION TITLE
Portable getch?

### DISCUSSION THREAD

**Discussion by andyl** [fuelen/owl#5] [General]
I’ve been unable to find a simple & portable `getch` implementation for Elixir. The use case is to detect single keypress events, without needing the return key.

Comparable tools like [inquirer/npm](https://www.npmjs.com/package/inquirer) and [tty/ruby](https://ttytoolkit.org/) use `getch` to build features like type-ahead, pull-down menus, etc.

To implement `getch` I've looked at ex_ncurses, rustler, System.cmd, Ports, etc. This is the best I’ve come up with so far:

```elixir
#!/usr/bin/env elixir

ruby_cmd = ~S"""
  ruby -e '
    require "io/console"
    @output = IO.new(4)
    @output.sync = true
    char = STDIN.getch
    @output.write(char)
  '
"""

IO.write("Press a key > ")
port = Port.open({:spawn, ruby_cmd}, [:binary, :nouse_stdio]) 
receive do
  {^port, {:data, result}} -> IO.puts("\nChar: #{result}")
end
```
This works on my Linux dev machine, but depends on Ruby.

Does anyone know of a portable `getch` implementation? Maybe with Rustler or precompiled binaries or ???

**Comment by fuelen** [comment]
I found this crate https://crates.io/crates/getch
it is quite small, we can play with it and Rustler and if it works okay, then implement as a separate package

**Comment by antedeguemon** [comment]
@fuelen @andyl I tried using Rustler + getch but couldn't get it to work. I think it is due to BEAM blocking the stdin port ([here](https://github.com/erlang/otp/blob/e8f26bba73949dc0eb8a49516a7191ffebfd6487/lib/kernel/src/user_drv.erl#L125)).

So I came with a dirty solution that closes the stdin port opened by the user driver and re-opens it on a process I can control:
https://gist.github.com/antedeguemon/565d65718a4625dcec2d83d51614a9c5

It works well but it has the big downside of overriding the IO port, meaning regular IO operations won't work. Here is an asciicast of the script:

[![asciicast](https://asciinema.org/a/mknz6bilhriB2nFk9laCHyFUb.svg)](https://asciinema.org/a/mknz6bilhriB2nFk9laCHyFUb)

**Comment by fuelen** [comment]
Hey @antedeguemon 

Cool, that's interesting. I went to Google to find what is tty_sl and quickly found this repo https://github.com/swalkerhppr/elixir-io_tty :)
Seems like manual stop of the port can be replaced with `-noinput` flag. @andyl have you seen the repo?

**Comment by andyl** [comment]
@antedeguemon thanks for posting your code I was able to run it locally.  Studying.

@fuelen - have not seen `elixir-io_tty` before - it looks interesting!

**Comment by andyl** [comment]
I had another look - seems like `tty_sl` is an Erlang module that makes it possible to read character-by-character, and is used in `IEx.CLI.start`.  I am unable to find any docs on `tty_sl`.  Anyone have a link for this?

**Comment by fuelen** [comment]
> Anyone have a link for this?

Unfortunately no

**Comment by andyl** [comment]
Here's a thread with some good resources:

https://groups.google.com/g/elixir-lang-talk/c/WtzbZcAk6XQ

**Comment by zachallaun** [comment]
Hi friends! I’ve been reading through the various linked resources from this thread this morning and wanted to follow up to see if any conclusions had been drawn. 

It looks to me like the key is:

```elixir
Port.open({:spawn, "tty_sl -c -e"}, [:binary, :eof])
```

which will send input characters to the opening process. The question then is what are the additional responsibilities that process has to keep everything working? Some misc things that I believe would need to be handled:

- Obvious stuff like writing characters back to the terminal (if desired)
- Handle cursor position
- Intercept output from other processes (register as group leader)

Other questions:

- How does the interact with IEx? E.g. running with `iex -S mix …` and hitting a `dbg`
- I’m sure there are other implications here…

**Comment by andyl** [comment]
Here's a [gist](https://gist.github.com/andyl/d4ebbf26cce9a18995e319bb997f5db3) for a 'getch' demo script.  It works for me on Ubuntu 22.04 with OTP25 and Elixir 1.15.  Give it a try!

**Reply by zachallaun** [reply]
Worked well on Ubuntu 20.04 via WSL, but only had a moment to test it. 

Things to try:
- Ctrl-c / Ctrl-g
- Starting with IEx, is there a way to hijack control from IEx/yield it back later?

**Comment by fuelen** [comment]
Also, here are some upcoming changes which may change the current state https://github.com/erlang/otp/pull/6144
I've just tried to run the script using Erlang from master (OTP 26) and the script prints the following error:
```
PRESS ANY KEY> sh: line 1: exec: tty_sl: not found
```

**Reply by zachallaun** [reply]
There may be some hints for what can be used on OTP 26 here: https://github.com/erlang/otp/blob/master/lib/kernel/src/prim_tty.erl

**Reply by andyl** [reply]
I get the same error `tty_sl: not found` using `erlang 26.0.2` and `elixir 1.15.4.otp-26` on Ubuntu 22.04.  Had a look at `prim_tty.erl` & not able to figure out a solution.  Anyone have a working example to share?

**Reply by zachallaun** [reply]
I'd love a "native" solution for OTP 24-26, but to be honest, I've lately been looking more towards Rustler/Zigler to wrap a more "native" getch solution based on termbox2 (C) or crossterm (rust). What I really, _really_ want is something that puts the term into raw mode only for handling `getch`. I think it's possible by implementing the IO protocol and making sure the appropriate stuff gets cached while a `getch` is pending, but I've not looked into the edge cases. 

Basically: I want to have my cake and eat it too 😂 I want to be able to just `puts` my output and keep things simple, but `getch` a single char to improve UX.

**Reply by andyl** [reply]
The [source](https://github.com/erlang/otp/blob/8d9828f0f07bae96a79d471b38b0b5ed8c2bd6c2/erts/emulator/nifs/common/prim_tty_nif.c) for the `prim_tty` nifs is more than 1000 lines. Probably somewhere in there is code that works like `tty_sl`, but there is no documentation.  A standalone `getch` implementation would be great.

**Comment by andyl** [comment]
This [issue](https://github.com/erlang/otp/issues/8037) has a demo of `getch` for Erlang/OTP26 that uses `prim_tty`...

**Comment by andyl** [comment]
OTP 28 is now live and has support for `raw mode`, giving a documented API for `getch`.  See the @zachallaun thread on [Elixir Forum](https://elixirforum.com/t/raw-terminal-mode-coming-to-otp-28/67491/16) for more info.
